### PR TITLE
New example: voltage clamp

### PR DESCRIPTION
**Pull request by AurelienJaquier**
## Description

New example: voltage clamp with data from channelpedia and jupyter notebook

## Checklist:
- [x] Unit tests are added to cover the changes (skip if not applicable).
- [x] The changes are mentioned in the documentation (skip if not applicable).
- [ ] CHANGELOG file is updated (skip if not applicable).

### CODE REVIEW

File: docs/source/eFeatures.rst [hunk 1]
```diff
@@ -2085,6 +2088,126 @@ with impedance_max_freq being a setting with 50.0 as a default value.
...
+    # correct time so that it starts from 0
+    time_interval -= time_interval[0]
+
+    # fit
+    popt, _ = curve_fit(
+        exp_fit,
+        time_interval,
+        voltage_interval,
+        p0=(1., voltage_interval[-1], voltage_interval[0] - voltage_interval[-1]),
+        bounds=(((0, -np.inf, -np.inf), (np.inf, np.inf, 0))),  # positive tau, negative A1
```
Comment (ilkilic) [inline docs/source/eFeatures.rst:2127]: is the extra parenthesis inside `bounds` necessary?

File: docs/source/eFeatures.rst [hunk 2]
```diff
@@ -2085,6 +2088,126 @@ with impedance_max_freq being a setting with 50.0 as a default value.
...
+    # correct time so that it starts from 0
+    time_interval -= time_interval[0]
+
+    # fit
+    popt, _ = curve_fit(
+        exp_fit,
+        time_interval,
+        voltage_interval,
+        p0=(1., voltage_interval[-1], max(0, voltage_interval[0] - voltage_interval[-1])),
+        bounds=(((0, -np.inf, 0), np.inf)),  # positive tau, positive A1
```
Comment (ilkilic) [inline docs/source/eFeatures.rst:2162]: wouldn't (np.inf, np.inf, np.inf) be more clear for the upper bound?

File: docs/source/eFeatures.rst [hunk 3]
```diff
@@ -2085,6 +2088,126 @@ with impedance_max_freq being a setting with 50.0 as a default value.
...
+    # correct time so that it starts from 0
+    time_interval -= time_interval[0]
+
+    # fit
+    popt, _ = curve_fit(
+        exp_fit,
+        time_interval,
+        voltage_interval,
+        p0=(1., voltage_interval[-1], voltage_interval[0] - voltage_interval[-1]),
+        bounds=(((0, -np.inf, 0), np.inf)),  # positive tau, positive A1
```
Comment (ilkilic) [inline docs/source/eFeatures.rst:2206]: same as before

File: efel/pyfeatures/pyfeatures.py [hunk 4]
```diff
@@ -352,3 +355,148 @@ def phaseslope_max() -> np.ndarray | None:
...
+    try:
+        popt, _ = curve_fit(
+            exp_fit,
+            time_interval,
+            voltage_interval,
+            p0=(1., voltage_interval[-1], voltage_interval[0] - voltage_interval[-1]),
+            bounds=(((0, -np.inf, 0), np.inf)),  # positive tau, positive A1
+            nan_policy="omit",
+        )
+    except (ValueError, RuntimeError, OptimizeWarning):
```
Comment (ilkilic) [inline efel/pyfeatures/pyfeatures.py:499]: I believe `OptimizeWarning` is a warning, not an exception, so I think it is not necessary to include it

File: examples/voltage_clamp/voltage_clamp.ipynb [hunk 5]
```diff
@@ -0,0 +1,801 @@
...
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Activation traces\n",
+    "\n",
+    "We will start to load the data form the nwb file. In this notebook, we will use the traces from thress experiments: Activation, Deactivation and Inactivation."
```
Comment (ilkilic) [inline examples/voltage_clamp/voltage_clamp.ipynb:44]: thress -> three ?

File: examples/voltage_clamp/voltage_clamp.ipynb [hunk 6]
```diff
@@ -0,0 +1,801 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
```
Comment (ilkilic) [inline examples/voltage_clamp/voltage_clamp.ipynb:55]: perhaps we could remove the warning

File: examples/voltage_clamp/voltage_clamp.ipynb [hunk 7]
```diff
@@ -0,0 +1,801 @@
...
+    "plt.ylabel(\"Steady State Current (nA)\")\n",
+    "plt.title(\"Activation steady state\")"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We can also plot the current base. Since this feature is compute on the trace before the stimulus is applied, it should be constent across the traces."
```
Comment (ilkilic) [inline examples/voltage_clamp/voltage_clamp.ipynb:305]: compute -> computed
constent -> constant

File: examples/voltage_clamp/voltage_clamp.ipynb [hunk 8]
```diff
@@ -0,0 +1,801 @@
...
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Deactivation traces\n",
+    "\n",
+    "Now we will also perform efeature extraction on traces from deactivation experiment. First, we have load the data from the file. Once again, we will put the current data under the 'V' key representing the voltage, so that efel treats our trace like any voltage trace. Notice that here, we are setting the stimulus start at 401 and stimulus end at 599 because from 400 to 600 is the interval of interest for our time constant fit, and we start slightly after and end slightly before 'real' stimulus in order to prevent artifacts that show when there is a stimulus change to affect our computation"
```
Comment (ilkilic) [inline examples/voltage_clamp/voltage_clamp.ipynb:401]: we have load -> we load
change to affect -> change that affect ?

File: examples/voltage_clamp/voltage_clamp.ipynb [hunk 9]
```diff
@@ -0,0 +1,801 @@
...
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Inactivation traces\n",
+    "\n",
+    "Now we will also perform efeature extraction on traces from inactivation experiment. First, we have load the data from the file. Once again, we will put the current data under the 'V' key representing the voltage, so that efel treats our trace like any voltage trace. For this one, we will have to extract the features in two times, because they depend on which stimulus interval we are interested in. We will first look into the features we can extract from the stimulus interval between 1600 and 1700 ms."
```
Comment (ilkilic) [inline examples/voltage_clamp/voltage_clamp.ipynb:524]: we have load -> we load
in two times -> in two steps ?

File: examples/voltage_clamp/voltage_clamp.ipynb
Comment (darshanmandge) [inline examples/voltage_clamp/voltage_clamp.ipynb:1]: `Warning: `product` is deprecated as of NumPy 1.25.0, and will be removed in NumPy 2.0. Please use `prod` instead.
  "dt": np.array(rep["x_interval"], dtype="float32"),`
  Are these warnings coming from Efel, and can we do something about them?

Comment (darshanmandge) [inline examples/voltage_clamp/voltage_clamp.ipynb:1]: > eFel usually performs data manipulation on voltage traces, but here with voltage clamp data, we have current traces. In order to use the eFel features with voltage clamp data, we have to give the current recording to the "V" key in the trace dict. eFel will treat the current trace as if it were a voltage trace.

Can you say something similar to: "eFEL was designed mainly to extract features from voltage timeseries. The voltage clamp data are currents. To use the eFEL features with voltage clamp data, we have to give the current recording to the "V" key in the trace dict. eFEL will treat the current trace as if it were a voltage trace." ? 
Also, `eFel` can be changed to `eFEL` elsewhere too.

File: docs/source/eFeatures.rst (hunk 1 above)
Comment (AurelienJaquier) [reply docs/source/eFeatures.rst]: Fixed in latest commit

File: docs/source/eFeatures.rst (hunk 2 above)
Comment (AurelienJaquier) [reply docs/source/eFeatures.rst]: I like it this way. The code is shorter and I don't think it is less clear

File: docs/source/eFeatures.rst (hunk 3 above)
Comment (AurelienJaquier) [reply docs/source/eFeatures.rst]: Fixed in latest commit

File: efel/pyfeatures/pyfeatures.py [hunk 10]
```diff
@@ -352,3 +355,148 @@ def phaseslope_max() -> np.ndarray | None:
...
+        popt, _ = curve_fit(
+            exp_fit,
+            time_interval,
+            voltage_interval,
+            p0=(
+                1., voltage_interval[-1], max(
+                    0, voltage_interval[0] - voltage_interval[-1]
+                )
+            ),
+            bounds=(((0, -np.inf, 0), np.inf)),  # positive tau, positive A1
```
[the remark this answers is not in the document]
Comment (AurelienJaquier) [reply efel/pyfeatures/pyfeatures.py]: Fixed in latest commit

File: examples/voltage_clamp/voltage_clamp.ipynb (hunk 5 above)
Comment (AurelienJaquier) [reply examples/voltage_clamp/voltage_clamp.ipynb]: Fixed in latest commit

File: examples/voltage_clamp/voltage_clamp.ipynb (hunk 6 above)
Comment (AurelienJaquier) [reply examples/voltage_clamp/voltage_clamp.ipynb]: Removed in latest commit

File: examples/voltage_clamp/voltage_clamp.ipynb (hunk 7 above)
Comment (AurelienJaquier) [reply examples/voltage_clamp/voltage_clamp.ipynb]: Fixed in latest commit

File: examples/voltage_clamp/voltage_clamp.ipynb (hunk 8 above)
Comment (AurelienJaquier) [reply examples/voltage_clamp/voltage_clamp.ipynb]: Fixed in latest commit

File: examples/voltage_clamp/voltage_clamp.ipynb (hunk 9 above)
Comment (AurelienJaquier) [reply examples/voltage_clamp/voltage_clamp.ipynb]: Fixed in latest commit

File: examples/voltage_clamp/voltage_clamp.ipynb
Comment (AurelienJaquier) [reply examples/voltage_clamp/voltage_clamp.ipynb:1]: It was from my local venv having old software. It is fixed now

Comment (AurelienJaquier) [reply examples/voltage_clamp/voltage_clamp.ipynb:1]: Done in latest commit

**Comment by AurelienJaquier** [comment]
I also added fitting curves to the plot of the example as requested by @darshanmandge in latest commit

File: examples/voltage_clamp/voltage_clamp.ipynb
Comment (darshanmandge) [inline examples/voltage_clamp/voltage_clamp.ipynb:1]: The inactivation steady-state curve could not be fit using the features from inactivation channel currents?

Comment (AurelienJaquier) [reply examples/voltage_clamp/voltage_clamp.ipynb:1]: Added in latest commit

**Review by darshanmandge — APPROVED** [review]
Thanks @AurelienJaquier !

**Review by ilkilic — APPROVED** [review]
Thank you @AurelienJaquier !

**Pull request merged**